Update to MLIR d14cfe10341681

mikeurbach opened the pull request:
This MLIR commit includes a small fix for ODS custom directives.

We also need some minor updates to work with the latest MLIR:

1. Fully qualified `cppNamespace` for dialects using that
2. Update custom ODS builders to remove the `Builder` and
   `OperationState` parameters, which are automatically inserted

teqdruid reviewed (approved):
Looks good to me. Pretty mechanical changes. Thanks for doing this!

mikeurbach commented:
@stephenneuendorffer @lattner @maerhart here's another MLIR update that touches most dialects. Please take a look when you get a chance.

Regarding the dialect namespaces: I see we have a few different patterns here. We have `::circt::firrtl`, `::circt::rtl`, and `::circt::sv`, none of which appear to be nested under the `mlir` namespace. But then we have `::mlir::llhd`, which is under `mlir`, but not `circt`. Finally, `Handshake` and `StaticLogic` have an empty string for their `cppNamespace`. It seems like it is worth unifying this at some point, but I didn't take the liberty in this PR. If now would be a good time to align the namespaces between dialects, I can address it here.

stephenneuendorffer commented:
LGTM.  I think there are still some problems with Tablegen'd ops that refer to Traits and Interfaces, but don't live in the mlir namespace.   There's a patch here https://reviews.llvm.org/D88251 that shows the problem.  I suspect this will cause problems with LLHD for instance.

stephenneuendorffer commented:
And In general, moving things into the circt namespace is a good thing...

maerhart commented:
I agree LLHD should be moved into the CIRCT namespace at some point. This is a remnant from before it was merged into CIRCT.

mikeurbach commented:
I can move things into the `circt` namespace on this PR. I'm looking at that issue you raised @stephenneuendorffer. What do you think is the best path forward for CIRCT? I think we could put the CIRCT namespace inside MLIR like `::mlir::circt`, which is what npcomp is doing [here](https://github.com/llvm/mlir-npcomp/commit/d62f8227c2a205931158a59afd0c0c4e1e15760a). Or would it be preferable to use `::circt` at the top level for the long run?

mikeurbach commented:
Ok, I moved `LLHD`, `Handshake`, and `StaticLogic` to generate their ODS operations under the `::circt` namespace:

```
~/circt$ grep -R cppNamespace include/                                                                                                                                                        
include/circt/Dialect/Handshake/HandshakeOps.td:  let cppNamespace = "::circt::handshake";                                                                                                                         
include/circt/Dialect/RTL/RTL.td:  let cppNamespace = "::circt::rtl";                                                                                                                                              
include/circt/Dialect/FIRRTL/FIRRTL.td:  let cppNamespace = "::circt::firrtl";                                                                                                                                     
include/circt/Dialect/LLHD/IR/LLHD.td:  let cppNamespace = "::circt::llhd";                                                                                                                                        
include/circt/Dialect/LLHD/IR/SignalOps.td:  let cppNamespace = "::circt::llhd";                                                                                                                                   
include/circt/Dialect/StaticLogic/StaticLogic.td:  let cppNamespace = "::circt::staticlogic";                                                                                                                      
include/circt/Dialect/SV/SV.td:  let cppNamespace = "::circt::sv";
```

In some cases with LLHD, I added `using namespace mlir;` to bridge the gap now that it is no longer under the `::mlir` namespace.

What do we think? Having everything under `::circt` and `using namespace mlir;` in the places we need it seems ok to me. This is what `FIRRTL`, `RTL`, and `SV` have been doing.

teqdruid commented:
::circt is what I prefer (since it's shorter). I don't want to change it
later on (from ::mlir::circt to ::circt) as that will break a bunch of code
so the change would have to touch a lot of code.

I moved from ::mlir::esi to ::circt::esi (just yesterday) and I ran into a
lot of missing symbols... Pretty easy/mechanical to fix in the headers, but
a real pain. I think you made the right call on using namespace ::mlir.


On Fri, Sep 25, 2020 at 11:36 AM mikeurbach <notifications@github.com>
wrote:

> Ok, I moved LLHD, Handshake, and StaticLogic to generate their ODS
> operations under the ::circt namespace:
>
> ~/circt$ grep -R cppNamespace include/
> include/circt/Dialect/Handshake/HandshakeOps.td:  let cppNamespace = "::circt::handshake";
> include/circt/Dialect/RTL/RTL.td:  let cppNamespace = "::circt::rtl";
> include/circt/Dialect/FIRRTL/FIRRTL.td:  let cppNamespace = "::circt::firrtl";
> include/circt/Dialect/LLHD/IR/LLHD.td:  let cppNamespace = "::circt::llhd";
> include/circt/Dialect/LLHD/IR/SignalOps.td:  let cppNamespace = "::circt::llhd";
> include/circt/Dialect/StaticLogic/StaticLogic.td:  let cppNamespace = "::circt::staticlogic";
> include/circt/Dialect/SV/SV.td:  let cppNamespace = "::circt::sv";
>
> In some cases with LLHD, I added using namespace mlir; to bridge the gap
> now that it is no longer under the ::mlir namespace.
>
> What do we think? Having everything under ::circt and using namespace
> mlir; in the places we need it seems ok to me. This is what FIRRTL, RTL,
> and SV have been doing.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/llvm/circt/pull/102#issuecomment-699088554>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALNXYHVGISB75FIUFTYZDTSHTPLTANCNFSM4RY6KIXQ>
> .
>

teqdruid commented on lib/Dialect/LLHD/IR/LLHDOps.cpp:
```diff
@@ -122,7 +123,7 @@ struct constant_int_all_ones_matcher {
 } // anonymous namespace
 
-unsigned mlir::llhd::getLLHDTypeWidth(Type type) {
+unsigned llhd::getLLHDTypeWidth(Type type) {
```
Should this be circt::llhd, for clarity?

mikeurbach replied:
Ah yep, I will keep the fully qualified `circt::llhd` namespaces where they used to be `mlir::llhd`.

teqdruid commented on lib/Dialect/LLHD/Transforms/ProcessLoweringPass.cpp:
```diff
@@ -126,7 +127,6 @@ void ProcessLoweringPass::runOnOperation() {
 
-std::unique_ptr<OperationPass<ModuleOp>>
-mlir::llhd::createProcessLoweringPass() {
+std::unique_ptr<OperationPass<ModuleOp>> llhd::createProcessLoweringPass() {
```
We should probably add a clang-format check to our automated tests like llvm.

teqdruid reviewed (approved):
nit: You've used ::circt::llhd in some places and just ::llhd in others.

teqdruid reviewed (approved):
Looks good! Thanks

lattner reviewed (approved):
This looks great!  I really appreciate your work on these merges, thank you!

mikeurbach commented:
No problem, glad to keep using the latest MLIR.
